### PR TITLE
always pack a few newest ancient slots

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Problem
skipping rewrites and ancient packing testing are illustrating some real world issues against mnb.
Specifically, the algorithm can get stuck only considering old slots and then failing on the same old slots each time it is called. This prevents the algorithm from making progress. A way to tweak this is to alter slot selection.

#### Summary of Changes
Always include some newest ancient slots so that we have fresh storages in the mix each time and high # target slots. This allows us to handle the multi ref newest ancient storages. This also results in us continuing to select the highest slot #s possible. This keeps the set of ancient storages at the highest slot #s, which helps overall performance, such as root checking.

Fixes #

### CODE REVIEW

File: accounts-db/src/ancient_append_vecs.rs
Comment (brooksprumo) [inline accounts-db/src/ancient_append_vecs.rs]: How about cloning only the max value, instead of all of them:

```suggestion
        let max_slot = slots.iter().max().cloned().unwrap_or_default();
```

File: accounts-db/src/ancient_append_vecs.rs
Comment (brooksprumo) [inline accounts-db/src/ancient_append_vecs.rs]: Can this `100` be a constant somewhere outside this function?

File: accounts-db/src/ancient_append_vecs.rs
Comment (brooksprumo) [inline accounts-db/src/ancient_append_vecs.rs]: Can this be `is_high_slot`?

```suggestion
    is_high_slot: bool,
```

(and the other places in the file too)

File: accounts-db/src/ancient_append_vecs.rs
Comment (brooksprumo) [inline accounts-db/src/ancient_append_vecs.rs]: With the new conditional there's now a mix of && and ||. Is the code in the PR what's desired?

Explicitly, this PR is:
```
(!info.high_slot && storages_remaining + ancient_storages_required < low_threshold)
                || ancient_storages_required as u64 > u64::from(tuning.max_resulting_storages)
```

Wasn't sure if it should be:
```
!info.high_slot &&
                (storages_remaining + ancient_storages_required < low_threshold || ancient_storages_required as u64 > u64::from(tuning.max_resulting_storages))
```

File: accounts-db/src/ancient_append_vecs.rs
Comment (brooksprumo) [inline accounts-db/src/ancient_append_vecs.rs]: By doing these highest slots first, is there a chance we use up all the "available target slots"?

File: accounts-db/src/ancient_append_vecs.rs
Comment (jeffwashington) [reply accounts-db/src/ancient_append_vecs.rs]: done

File: accounts-db/src/ancient_append_vecs.rs
Comment (jeffwashington) [reply accounts-db/src/ancient_append_vecs.rs]: yes, i knew that was bad :(

File: accounts-db/src/ancient_append_vecs.rs
Comment (jeffwashington) [reply accounts-db/src/ancient_append_vecs.rs]: done

File: accounts-db/src/ancient_append_vecs.rs
Comment (jeffwashington) [reply accounts-db/src/ancient_append_vecs.rs]: the truncate change determines where we draw the line for pack and ignore. It is possible to use up all the target slots. But we are picking 100 new slots, 10(?) ideal output slots, and 130MB ideal sizes. The 100 newest slots shouldn't be excessively large. They should be normal validator steady state behaivor.

File: accounts-db/src/ancient_append_vecs.rs
Comment (brooksprumo) [inline accounts-db/src/ancient_append_vecs.rs]: nit:
```suggestion
                .then_with(|| r.should_shrink.cmp(&l.should_shrink))
```

**Review by brooksprumo — COMMENTED** [review]
Code looks good. Looking into testing now. Do you want tests in this PR before merging, or afterwards?

**Review by HaoranYi — DISMISSED** [review]
lgtm!

I think we may change the first sort of the candidate ancient slots to pack by bytes saved to by dirty ratio.

`self.sort_shrink_indexes_by_bytes_saved();`

sort by `bytes_saved` will more likely pickup big storage, while sort by `dirty ratio` will more likely pickup small storages.

File: accounts-db/src/ancient_append_vecs.rs
Comment (jeffwashington) [reply accounts-db/src/ancient_append_vecs.rs]: done

**Comment by jeffwashington** [comment]
> Do you want tests in this PR before merging

before. this is enabled by default, even though it shouldn't ever run unless you skip rewrites.

**Comment by jeffwashington** [comment]
> lgtm!
> 
> I think we may change the first sort of the candidate ancient slots to pack by bytes saved to by dirty ratio.
> 
> `self.sort_shrink_indexes_by_bytes_saved();`
> 
> sort by `bytes_saved` will more likely pickup big storage, while sort by `dirty ratio` will more likely pickup small storages.

this was on purpose. maybe misguided. but, idea is we want to focus on re-packing (ie. shrinking + combining with other alive ancient accounts) the slots with the most byte savings. And, yes, they may be large. We stop when we've scheduled 55% of the total dead bytes (I think). Otherwise, we might never pack the largest storages with the most dead accounts. Those storages save us the most bytes on disk.

**Review by brooksprumo — APPROVED** [review]
Lgtm. But I also was the one that wrote the tests, so it'd be good to have another pair of eyes review that part.

**Review by HaoranYi — APPROVED** [review]
lgtm

**Pull request merged**
